src/zkevm_specs/evm/step.py

```diff
@@ -0,0 +1,432 @@
+        is_persistent: int,
+        rw_counter_end_of_reversion: int,
+        r: int,
+    ):
+        caller_new_balance, caller_prev_balance = self.w_lookup(
+            RWTableTag.AccountBalance, [caller_address], is_persistent, rw_counter_end_of_reversion)[:2]
+        callee_new_balance, callee_prev_balance = self.w_lookup(
+            RWTableTag.AccountBalance, [callee_address], is_persistent, rw_counter_end_of_reversion)[:2]
+
+        # Verify caller's new balance is subtracted by value and not underflow
+        bytes_caller_prev_balance = self.decompress(caller_prev_balance, 32, r)
+        bytes_caller_new_balance = self.decompress(caller_new_balance, 32, r)
+        caller_carries = self.allocate_bool(32)
+        assert_addition(bytes_caller_new_balance, bytes_value, bytes_caller_prev_balance, caller_carries)
+        assert caller_carries[31] == 0
+
+        # Verify callee's new balance is added by value and not overflow
+        bytes_callee_prev_balance = self.decompress(callee_prev_balance, 32, r)
+        bytes_callee_new_balance = self.decompress(callee_new_balance, 32, r)
+        callee_carries = self.allocate_bool(32)
+        assert_addition(bytes_callee_prev_balance, bytes_value, bytes_callee_new_balance, callee_carries)
+        assert callee_carries[31] == 0
+
+    def assert_memory_expansion(
+        self,
+        bytes_cd_offset: Sequence[int],
+        bytes_cd_length: Sequence[int],
+        bytes_rd_offset: Sequence[int],
+        bytes_rd_length: Sequence[int],
+    ) -> Tuple[int, int]:
+        cd_offset = le_to_int(bytes_cd_offset[:5])
+        cd_length = le_to_int(bytes_cd_length)
+        rd_offset = le_to_int(bytes_rd_offset[:5])
+        rd_length = le_to_int(bytes_rd_length)
+
+        next_memory_size = self.allocate(1)[0]
+
+        has_cd_length = not self.is_zero(cd_length)
+        has_rd_length = not self.is_zero(rd_length)
+        bytes_next_memory_size_cd = self.allocate_byte(4)
+        bytes_next_memory_size_rd = self.allocate_byte(4)
+        next_memory_size_cd = has_cd_length * le_to_int(bytes_next_memory_size_cd)
+        next_memory_size_rd = has_rd_length * le_to_int(bytes_next_memory_size_rd)
+
+        # Verify next_memory_size_cd is correct
+        if has_cd_length:
+            assert sum(bytes_cd_offset[5:]) == 0
+        self.fixed_lookup(FixedTableTag.Range32, [32 * next_memory_size_cd - has_cd_length * (cd_offset + cd_length)])
+
+        # Verify next_memory_size_rd is correct
+        if has_rd_length:
+            assert sum(bytes_rd_offset[5:]) == 0
+        self.fixed_lookup(FixedTableTag.Range32, [32 * next_memory_size_rd - has_rd_length * (rd_offset + rd_length)])
+
+        # Verify next_memory_size == \
+        #   max(self.call.memory_size, next_memory_size_cd, next_memory_size_rd)
+        assert next_memory_size in [self.call.memory_size, next_memory_size_cd, next_memory_size_rd]
+        self.bytes_range_lookup(next_memory_size - self.call.memory_size, 4)
```

Comment:
Why do we use 4 bytes here while cap the length at 5 bytes? We should also use variables to represent these constants to make them more meaningful and consistent.

Reply:
Agree, I should mention why we store memory size in word and why it should fit in 4 bytes.

The reason to store memory size in word is because memory expansion always expands to multiple of word, and it makes the quadratic gas cost easier to calculate (we don't need to divide it by 32 everytime).

The reason to use 4 bytes is still due to the quadratic gas cost, the memory size in word could be up to <code>2<sup>32</sup>-1</code>, which doesn't overflow `u64` when doing square (mimicking the behavior of current EVM).

Reply:
Got it.